### PR TITLE
driver/pca9633: add new driver

### PR DESCRIPTION
**Pull request by HendrikVE**
### Contribution description

This PR adds support for the NXP PCA9633. The PCA9633 is an I2C-bus controlled 4-bit LED driver optimized for Red/Green/Blue/Amber (RGBA) color mixing applications.

### Testing procedure

The given (manual) test provides single commands for every driver function in addition to the option for a complete demo run.

### CODE REVIEW

File: drivers/pca9633/pca9633.c [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+ * @return  0 on success
+ * @return  -PCA9633_ERROR_I2C if acquiring of I2C bus fails
+ * @return  -EIO When slave device doesn't ACK the byte
+ * @return  -ENXIO When no devices respond on the address sent on the bus
+ * @return  -ETIMEDOUT When timeout occurs before device's response
+ * @return  -EINVAL When an invalid argument is given
+ * @return  -EOPNOTSUPP When MCU driver doesn't support the flag operation
+ * @return  -EAGAIN When a lost bus arbitration occurs
+ */
+int _read_reg(pca9633_t* dev, uint8_t reg, uint8_t* data);
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:58]:
```suggestion
static int _read_reg(pca9633_t* dev, uint8_t reg, uint8_t* data);
```

File: drivers/pca9633/pca9633.c [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+ * @return  0 on success
+ * @return  -PCA9633_ERROR_I2C if acquiring of I2C bus fails
+ * @return  -EIO When slave device doesn't ACK the byte
+ * @return  -ENXIO When no devices respond on the address sent on the bus
+ * @return  -ETIMEDOUT When timeout occurs before device's response
+ * @return  -EINVAL When an invalid argument is given
+ * @return  -EOPNOTSUPP When MCU driver doesn't support the flag operation
+ * @return  -EAGAIN When a lost bus arbitration occurs
+ */
+int _write_reg(pca9633_t* dev, uint8_t reg, uint8_t data);
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:40]:
```suggestion
static int _write_reg(pca9633_t* dev, uint8_t reg, uint8_t data);
```

File: drivers/include/pca9633.h [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+ * @brief Mode register 2
+ */
+#define PCA9633_REG_MODE2       0x01
+
+/**
+ * @brief Brightness control LED0
+ */
+#define PCA9633_REG_PWM0        0x02
+
+/**
```
Comment (benpicco) [inline drivers/include/pca9633.h:48]: Move all those register defines to `drivers/pca9633/include/pca9633_registers.h`, no need to have them in the global API space.

File: drivers/pca9633/pca9633.c [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+    _write_reg(dev, PCA9633_REG_LEDOUT, new_reg);
+}
+
+void pca9633_set_ldr_state_all(pca9633_t* dev, uint8_t state)
+{
+    uint8_t new_reg = ( (state << PCA9633_BIT_LDR3)
+                       | (state << PCA9633_BIT_LDR2)
+                       | (state << PCA9633_BIT_LDR1)
+                       | (state << PCA9633_BIT_LDR0) );
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:174]:
```suggestion
    uint8_t new_reg = (state << PCA9633_BIT_LDR3)
                    | (state << PCA9633_BIT_LDR2)
                    | (state << PCA9633_BIT_LDR1)
                    | (state << PCA9633_BIT_LDR0);
```

excessive parenthesis

File: drivers/pca9633/pca9633.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    _write_reg(dev, reg_pwm, pwm);
+}
+
+void pca9633_set_grp_pwm(pca9633_t* dev, uint8_t pwm)
+{
+    _write_reg(dev, PCA9633_REG_GRPPWM, pwm);
+}
+
+void pca9633_set_blinking(pca9633_t* dev, uint8_t blink_period,
+        float on_off_ratio)
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:123]:
```suggestion
void pca9633_set_blinking(pca9633_t* dev, uint8_t blink_period,
        uint8_t on_off_ratio)
```

`float` is only available on Cortex-M4F and above - no need to pull in soft float libraries when you just want to use a value between 0 and 255.

File: drivers/pca9633/pca9633.c [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    }
+}
+
+void pca9633_set_ldr_state(pca9633_t* dev, uint8_t state, uint8_t ldr_bit)
+{
+    uint8_t prev_reg;
+    _read_reg(dev, PCA9633_REG_LEDOUT, &prev_reg);
+    uint8_t new_reg;
+
+    // first clear both bits of ldr
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:160]:
```suggestion
    /* first clear both bits of ldr */
```

RIOT coding style

File: drivers/pca9633/pca9633.c [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    if (dev->params.has_amber_channel) {
+        pca9633_set_pwm(dev, dev->params.reg_pwm_amber, a);
+    }
+}
+
+void pca9633_set_ldr_state(pca9633_t* dev, uint8_t state, uint8_t ldr_bit)
+{
+    uint8_t prev_reg;
+    _read_reg(dev, PCA9633_REG_LEDOUT, &prev_reg);
+    uint8_t new_reg;
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:158]: Why use two variables here?

File: drivers/include/pca9633.h (hunk 3 above)
Comment (HendrikVE) [reply drivers/include/pca9633.h:48]: E.g. `PCA9633_REG_PWM0` to `PCA9633_REG_PWM3` are needed when calling `pca9633_set_pwm` (see `tests/driver_pca9633/main.c` on line 90). Should they still be declared in `pca9633_regs.h` or in `pca9633.h`? It's my first driver in RIOT ;)

File: drivers/pca9633/pca9633.c (hunk 5 above)
Comment (HendrikVE) [reply drivers/pca9633/pca9633.c]: I changed it to an uint8_t

File: drivers/include/pca9633.h (hunk 3 above)
Comment (benpicco) [reply drivers/include/pca9633.h]: Usually it's a bit odd for an API to let the user write registers directly, but for writing PWM channels that's simple enough.

You might add some type-safety (and ease discoverability) by making the register definitions an `enum` instead of `#defines`

```C
typedef enum {
    PCA9633_REG_PWM0 = 0x2,
    PCA9633_REG_PWM1 = 0x3,
    …
} pca9633_reg_t;
```

The you would have

```C
void pca9633_set_pwm(pca9633_t* dev, pca9633_reg_t reg_pwm, uint8_t pwm)
```

Although it would make for an even easier to use API if `pca9633_set_pwm()` would simply take channel 0…3 as parameter.

Unfortunately we don't have a common API for external PWM devices yet, but it would still be a good advice to follow the API in `drivers/include/periph/pwm.h`.

File: drivers/pca9633/pca9633.c [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+            enabled = false;
+            bit0 = false;
+            bit1 = false;
+            break;
+    }
+
+    uint8_t new_reg = ( (enabled << PCA9633_BIT_AI2)
+                       | (bit1 << PCA9633_BIT_AI1)
+                       | (bit0 << PCA9633_BIT_AI0) );
+
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:211]: still excessive parenthesizes

File: drivers/pca9633/pca9633.c [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+void pca9633_set_auto_increment(pca9633_t* dev, uint8_t option)
+{
+    bool enabled, bit0, bit1;
+
+    switch (option) {
+
+        case PCA9633_AI_ALL:
+            enabled = true;
+            bit0 = false;
+            bit1 = false;
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:179]: Actually you could just do
```suggestion
            new_reg = (1 << PCA9633_BIT_AI2)
                    | (0 << PCA9633_BIT_AI1)
                    | (0 << PCA9633_BIT_AI0) );
```

File: drivers/pca9633/pca9633.c [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+            break;
+
+        case PCA9633_AI_IND:
+            enabled = true;
+            bit0 = false;
+            bit1 = true;
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:185]:
```suggestion
            new_reg = (1 << PCA9633_BIT_AI2)
                    | (0 << PCA9633_BIT_AI1)
                    | (1 << PCA9633_BIT_AI0);
```

(or leave out the 0 bits)

File: drivers/include/pca9633.h [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+ * @brief Control mode for blinking
+ */
+#define PCA9633_GROUP_CONTROL_MODE_BLINKING 0
+
+/**
+ * @brief Control mode for dimming
+ */
+#define PCA9633_GROUP_CONTROL_MODE_DIMMING  1
```
Comment (benpicco) [inline drivers/include/pca9633.h:101]: Better use an `enum` here so the API is more clear.

**Comment by HendrikVE** [comment]
Thank you @benpicco  for your advices. Like that?

File: drivers/include/pca9633.h [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+ * ((1 / 24 Hz) * 12 cycles)
+ */
+#define PCA9633_BLINKING_PERIOD_500_MS  12
+
+/**
+ * @brief Blinking period with a duration of 1 s
+ *
+ * ((1 / 24 Hz) * 24 cycles)
+ */
+#define PCA9633_BLINKING_PERIOD_1_S     24
```
Comment (benpicco) [inline drivers/include/pca9633.h:60]:
```suggestion
/**
 * @brief Blinking period with a duration of n ms
 *
 *  ms = ((1 / 24 Hz) * cycles)
 */
#define PCA9633_BLINKING_PERIOD(n) ((24 * (ms)) / 1000)
```

Comment (HendrikVE) [reply drivers/include/pca9633.h]: I chose the latter option.

File: drivers/include/pca9633.h [hunk 13]
```diff
@@ -30,39 +30,10 @@ extern "C"
...
- *
- * ((1 / 24 Hz) * 24 cycles)
- */
-#define PCA9633_BLINKING_PERIOD_1_S     24
-
 /**
  * @brief Blinking period with a maximum duration of ~10.73 s
  */
-#define PCA9633_BLINKING_PERIOD_MAX     255
+#define PCA9633_BLINKING_PERIOD_MAX     (65535)
```
Comment (benpicco) [inline drivers/include/pca9633.h:36]: Shouldn't that be 10625?

Comment (HendrikVE) [reply drivers/include/pca9633.h:36]: It doesn't matter. I simply took the largest value possible such that the if-condition in
```
uint16_t blink_period = (24 * blink_period_ms) / 1000;
if (blink_period > 255) {
    blink_period = 255;
}
```
is triggered and the value will be capped to 255 for the register value anyway.

Comment (HendrikVE) [reply drivers/include/pca9633.h:36]: But if you think its confusing I can change it.

Comment (benpicco) [reply drivers/include/pca9633.h:36]: It *is* confusing.
Since this is not used anywhere, it will only lead unsuspecting users astray ;)

If you change it, also better change it to `PCA9633_BLINKING_PERIOD_MAX_MS` so it's apparent what unit that number is in.

File: drivers/include/pca9633.h [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+{
+#endif
+
+#include "stdbool.h"
+#include "periph/i2c.h"
+
+#include "pca9633_regs.h"
+
+/**
+ * @brief Blinking period with a maximum duration of ~10.73 s
```
Comment (benpicco) [inline drivers/include/pca9633.h:34]: nitpick: 10625 ms is actually ~10.63 s ;)

File: drivers/include/pca9633.h (hunk 13 above)
Comment (HendrikVE) [reply drivers/include/pca9633.h]: ok^^ I changed it and I also mentioned the define in the function's documentation now.

File: drivers/include/pca9633.h [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+ * @param[in] blink_period_ms   Period in ms for one blink (turning off and on).
+ *                              Maximum period possible is
+ *                              PCA9633_BLINKING_PERIOD_MAX_MS ≈ 10.73 s. All
+ *                              values above this maximum will we capped to it.
+ * @param[in] on_off_ratio      Value between 0 and 255, where e.g. a value of
+ *                              64 (255/4) means 1/4 of the time the LEDs are on
+ *                              and 3/4 of the time the LEDs are off.
+ */
+void pca9633_set_blinking(pca9633_t* dev, uint16_t blink_period_ms,
+        uint8_t on_off_ratio);
```
Comment (benpicco) [inline drivers/include/pca9633.h:254]:
```suggestion
void pca9633_set_blinking(pca9633_t* dev, uint16_t blink_period_ms,
                          uint8_t on_off_ratio);
```

File: drivers/include/pca9633.h (hunk 14 above)
Comment (HendrikVE) [reply drivers/include/pca9633.h:34]: `10.73 s` is actually  the value I got from the data sheet. Let me check.

File: drivers/include/pca9633.h [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+ *          - PCA9633_LDR_STATE_ON
+ *          - PCA9633_LDR_STATE_IND
+ *          - PCA9633_LDR_STATE_IND_GRP
+ *
+ * @param[in] dev       Device descriptor of the PCA9633
+ * @param[in] state     One of the four possible states
+ * @param[in] ldr_bit   Lower bit of LDR* (see BIT_LDR*)
+ */
+void pca9633_set_ldr_state(pca9633_t* dev,
+        pca9633_ldr_state_t state, pca9633_pwm_channel_t pwm_channel);
```
Comment (benpicco) [inline drivers/include/pca9633.h:290]:
```suggestion
void pca9633_set_ldr_state(pca9633_t* dev,
                           pca9633_ldr_state_t state, pca9633_pwm_channel_t pwm_channel);
```

File: drivers/include/pca9633.h [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+ * @brief Set the group control mode.
+ *        There are two modes:
+ *          - PCA9633_GROUP_CONTROL_MODE_BLINKING
+ *          - PCA9633_GROUP_CONTROL_MODE_DIMMING
+ *
+ * @param[in] dev       Device descriptor of the PCA9633
+ * @param[in] mode      One of the two possible modes
+ */
+void pca9633_set_group_control_mode(pca9633_t* dev,
+        pca9633_group_control_mode_t mode);
```
Comment (benpicco) [inline drivers/include/pca9633.h:329]:
```suggestion
void pca9633_set_group_control_mode(pca9633_t* dev,
                                    pca9633_group_control_mode_t mode);
```

File: drivers/include/pca9633.h (hunk 14 above)
Comment (HendrikVE) [reply drivers/include/pca9633.h:34]:
```
A fifth 8-bit resolution (256 steps) Group PWM controller
has both a fixed frequency of 190 Hz and an adjustable frequency between 24 Hz to once
every 10.73 seconds with a duty cycle that is adjustable from 0 % to 99.6 % that is used to
either dim or blink all LEDs with the same value.
```

File: drivers/pca9633/pca9633.c [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+    int rc = _write_reg(dev, PCA9633_REG_MODE1, 0x0);
+    _write_reg(dev, PCA9633_REG_MODE2, 0x0);
+
+    if (rc != PCA9633_OK) {
+        return rc;
+    }
+
+    pca9633_set_ldr_state_all(dev, PCA9633_LDR_STATE_IND_GRP);
+    pca9633_set_group_control_mode(dev, PCA9633_GROUP_CONTROL_MODE_DIMMING);
+    pca9633_set_rgb(dev, 255, 255, 255);
```
Comment (benpicco) [inline drivers/pca9633/pca9633.c:79]: Does that enable the output on init?
I think that's not what most users would want - only enable the output when the user has requested it.

File: drivers/include/pca9633.h (hunk 14 above)
Comment (HendrikVE) [reply drivers/include/pca9633.h:34]: The 10.73 appears several times so I would conclude it's not a type there. Probably that's just the value with inaccuracies in mind.

File: drivers/pca9633/pca9633.c (hunk 18 above)
Comment (HendrikVE) [reply drivers/pca9633/pca9633.c:79]: It's turned on by default. The lines
```
pca9633_set_ldr_state_all(dev, PCA9633_LDR_STATE_IND_GRP);
pca9633_set_group_control_mode(dev, PCA9633_GROUP_CONTROL_MODE_DIMMING);
```
are somewhat of an initialization, so they should be kept. But I could remove
```
pca9633_set_rgb(dev, 255, 255, 255);
```
and exchange it with a call of `pca9633_turn_off()` if that's how the drivers normally operate.

Comment (benpicco) [reply drivers/pca9633/pca9633.c:79]: Yea you don't want the LED to flash each time the device is turned on.
And one could connect something else entirely to those PWM outputs, like a motor or such.
Generally, you don't want something to happen without the user explicitly calling a function to make it happen.

Comment (HendrikVE) [reply drivers/pca9633/pca9633.c:79]: Ah I mixed this in my head with another driver here. I had another look to the data sheet. For this one, there is no global `turned off` state. All LDRs (4 different for the pwm channels) are set separatly (`PCA9633_LDR_STATE_OFF` by default).

Because we do `pca9633_set_group_control_mode(dev, PCA9633_GROUP_CONTROL_MODE_DIMMING);` we don't need `pca9633_set_group_control_mode(dev, PCA9633_GROUP_CONTROL_MODE_DIMMING);`

Long story short: I remove all three lines and no need for `pca9633_turn_off` :D

**Comment by HendrikVE** [comment]
Thank you very much @benpicco for your detailed review!
